### PR TITLE
Add LaCrosse WS7000 support

### PR DESCRIPTION
**Pull request by zuckschwerdt**
See #1029

### CODE REVIEW

**Comment by zuckschwerdt** [comment]
Needs tidying up and some abstractions. WIP for now, but basically works.

File: src/devices/lacrosse_ws7000.c [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+                NULL);
+        /* clang-format on */
+
+        decoder_output_data(decoder, data);
+        return 1;
+    }
+    else if (type == 4) {
+        // 4 = WS7000-20 Thermo/Humidity/Barometer sensor
+        int sign           = b[1] & 0x8 ? -1 : 1;
+        double temperature = ((b[4] * 10) + (b[3] * 1) + (b[2] * 0.1)) * sign;
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: float is enough.

File: src/devices/lacrosse_ws7000.c [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        return 0;
+
+    unsigned max_bits = MIN(14 * 5, bitbuffer->bits_per_row[0] - start_pos);
+    unsigned len = unstuff_5to4(bitbuffer->bb[0], start_pos, max_bits, b);
+    if (len < 2) // at least type/addr nibbles
+        return 0;
+
+    //bitrow_printf(b, len * 8, "LaCrosse WS7000: ");
+
+    reflect_nibbles(b, len);
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: Before the reflection we should reject based on bit length. So far we know  4 types of sensors and they all have different lengths. So it should be cheap to reject here.

File: src/devices/lacrosse_ws7000.c [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+  - 3 = WS7000-15 Wind sensor ( interval 169s - Adr * 0.5s )
+  - 4 = WS7000-20 Thermo/Humidity/Barometer sensor ( interval 165s - Adr * 0.5s )
+  - 5 = WS2500-19 Brightness sensor ( interval 161s - Adr * 0.5s )
+- Address:  Value 0..7 for the sensor address
+    In case of a negative temperature the MSB of the Address becomes "1"
+- Data:     Nibbles with BCD encoded sensor data values.
+- XOR:      Nibble holding XOR of the S ^ A ^ Data-nibbles
+- Checksum: Sum of all nibbles + 5 ( S + A + nibble(0) + .. + nibble(n) + XOR + 5 ) & 0xF
+
+*/
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: Maybe if we could add a link to the documentation here? Or maybe just add that to the tests repo text description.

File: src/devices/lacrosse_ws7000.c [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    if (((add_bytes(b, len - 1) + 5) & 0xf) != b[len - 1]) {
+        //if (decoder->verbose > 1)
+            fprintf(stderr, "LaCrosse-WS7000: checksum error (add)\n");
+        return 0;
+    }
+
+    if (type == 0) {
+        // 0 = WS7000-27/28 Thermo sensor
+        int sign           = b[1] & 0x8 ? -1 : 1;
+        double temperature = ((b[4] * 10) + (b[3] * 1) + (b[2] * 0.1)) * sign;
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: float

File: src/devices/lacrosse_ws7000.c [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                NULL);
+        /* clang-format on */
+
+        decoder_output_data(decoder, data);
+        return 1;
+    }
+    else if (type == 1) {
+        // 1 = WS7000-22/25 Thermo/Humidity sensor
+        int sign           = b[1] & 0x8 ? -1 : 1;
+        double temperature = ((b[4] * 10) + (b[3] * 1) + (b[2] * 0.1)) * sign;
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: float

File: src/devices/lacrosse_ws7000.c [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        /* clang-format on */
+
+        decoder_output_data(decoder, data);
+        return 1;
+    }
+    else if (type == 3) {
+        // 3 = WS7000-15 Wind sensor
+        double speed     = (b[4] * 10) + (b[3] * 1) + (b[2] * 0.1);
+        double direction = ((b[7] >> 2) * 100) + (b[6] * 10) + (b[5] * 1);
+        double deviation = (b[7] & 0x3) * 22.5;
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: float

File: src/devices/lacrosse_ws7000.c [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        int brightness = (b[4] * 100) + (b[3] * 10) + (b[2] * 1);
+        int exposition = (b[7] * 100) + (b[6] * 10) + (b[5] * 1);
+
+        /* clang-format off */
+        data = data_make(
+                "model",            "",                 DATA_STRING, "LaCrosse-WS2500-19",
+                "id",               "",                 DATA_INT,    id,
+                "channel",          "",                 DATA_INT,    addr,
+                "light_lux",        "Brightness",       DATA_INT,    brightness * 100,
+                "light_min",        "Exposition",       DATA_INT,    exposition,
```
Comment (merbanan) [inline src/devices/lacrosse_ws7000.c]: I think we need to elaborate that min==minutes here.

Comment (merbanan) [reply src/devices/lacrosse_ws7000.c]: And maybe it should be exposure?

**Comment by merbanan** [comment]
Looks good, commit at will.

**Comment by ReMiOS** [comment]
Great Work !!
the lux value on the WS2500-19 is working already. only the minutes needs another nibble i think 
899 minutes gives 992 min (2 is from the exp nibble)

**Comment by ReMiOS** [comment]
time      : 2019-04-02 21:48:41
model     : LaCrosse-WS7000-15                     id        : 55
channel   : 7            Wind speed: 12.500        Wind direction: 200.000   Wind deviation: 0.000     MIC       : CHECKSUM

time      : 2019-04-02 21:44:10
model     : LaCrosse-WS2500-19                     id        : 87
channel   : 7            Brightness: 52800         Exposition: 992           MIC       : CHECKSUM


time      : 2019-04-02 21:50:15
model     : LaCrosse-WS7000-16                     id        : 39
channel   : 7            Rain counter: 858.600     MIC       : CHECKSUM

**Comment by zuckschwerdt** [comment]
@ReMiOS oh, i was actually wondering what that "2" is and why that one example didn't match :)
@merbanan thanks for the pointers. I'll also need to think further on the nibble vs byte arrangement and likely put the helper functions in util.c

**Comment by ReMiOS** [comment]
Great ! 
These are working fine already :)

time      : 2019-04-02 21:55:44
model     : LaCrosse-WS7000-20                     id        : 71
channel   : 7            Temperature: -25.400      Humidity  : 47            Pressure  : 995           MIC       : CHECKSUM

time      : 2019-04-02 21:56:23
model     : LaCrosse-WS7000-22/25                  id        : 23
channel   : 7            Temperature: -25.400      Humidity  : 47            MIC       : CHECKSUM

time      : 2019-04-02 21:57:30
model     : LaCrosse-WS7000-27/28                  id        : 7
channel   : 7            Temperature: -25.300      MIC       : CHECKSUM

**Comment by zuckschwerdt** [comment]
I got to ask -25 C? Where is that sensor placed?

**Comment by ReMiOS** [comment]
LOL ... freezer ;)

**Comment by merbanan** [comment]
For the record optimal freezer temperatures are less then -18. But colder then that does not prolong food durability. ;)

**Comment by zuckschwerdt** [comment]
@merbanan I'm still experimenting with the idea to formalize the signal and payload description in such a way that I can render out pretty documentation like in that docx.
I should be not to verbose with little markup, human readable and fit the markdown block.

**Comment by ReMiOS** [comment]
Update: The WS2500-19 is now also working :)

**Comment by zuckschwerdt** [comment]
The ID is type*16+addr. That way you can easily track each device even with address collision by one short key.

**Comment by ReMiOS** [comment]
Great ! now added to the master branch 
Thanks for your support !

**Comment by zuckschwerdt** [comment]
Happy to help. Have fun with your freezer sensors ;)

**Pull request merged**